go/oasis-node/cmd/common/metrics/metrics.go

```diff
@@ -179,56 +313,533 @@ func (s *pushService) worker() {
+	return nil
+}
+
+// updateCPU updates current memory usage attributes.
+func (c *cpuService) updateCPU() error {
+	c.Lock()
+	defer c.Unlock()
+
+	/// Obtain process CPU info.
+	proc, err := procfs.NewProc(c.pid)
+	if err != nil {
+		return errors.Wrap(err, fmt.Sprintf("CPU metric: failed to obtain proc object for PID %d", c.pid))
+	}
+	procStat, err := proc.Stat()
+	if err != nil {
+		return errors.Wrap(err, fmt.Sprintf("CPU metric: failed to obtain procStat object %d", c.pid))
+	}
+
+	c.utimeGauge.Set(float64(procStat.UTime) / float64(ClockTicks))
+	c.stimeGauge.Set(float64(procStat.STime) / float64(ClockTicks))
+
+	return nil
+}
+
+func (c *cpuService) worker() {
+	t := time.NewTicker(c.interval)
+	defer t.Stop()
+
+	for {
+		select {
+		case <-c.Quit():
+			break
+		case <-t.C:
+		}
+
+		if err := c.updateCPU(); err != nil {
+			c.Logger.Warn(err.Error())
+		}
+	}
+}
+
+// NewCPUService constructs a new memory usage service.
+//
+// This service will read CPU spent time info from process Stat file every
+// --metric.push.interval seconds.
+func NewCPUService() (service.BackgroundService, error) {
+	cs := &cpuService{
+		BaseBackgroundService: *service.NewBaseBackgroundService("cpu"),
+		pid:                   os.Getpid(),
+		interval:              viper.GetDuration(CfgMetricsPushInterval),
+
+		utimeGauge: prometheus.NewGauge(
+			prometheus.GaugeOpts{
+				Name: "oasis_worker_cpu_utime_seconds",
+				Help: "CPU user time spent by worker",
+			},
+		),
+
+		stimeGauge: prometheus.NewGauge(
+			prometheus.GaugeOpts{
+				Name: "oasis_worker_cpu_stime_seconds",
+				Help: "CPU system time spent by worker",
+			},
+		),
+	}
+
+	cpuCollectors := []prometheus.Collector{
+		cs.utimeGauge,
+		cs.stimeGauge,
+	}
+
+	// CPU metrics are singletons per process. Ensure to register them only once.
+	cpuServiceOnce.Do(func() {
+		prometheus.MustRegister(cpuCollectors...)
+	})
+
+	return cs, nil
+}
+
+type netService struct {
+	service.BaseBackgroundService
+	sync.Mutex
+
+	interval time.Duration
+
+	ReceiveBytesGauge    prometheus.Gauge
+	ReceivePacketsGauge  prometheus.Gauge
+	TransmitBytesGauge   prometheus.Gauge
+	TransmitPacketsGauge prometheus.Gauge
+}
+
+func (n *netService) Start() error {
+	go n.worker()
+	return nil
+}
+
+// updateNetwork updates current network interface statistics.
+func (n *netService) updateNetwork() error {
+	n.Lock()
+	defer n.Unlock()
+
+	// Obtain process Network info.
+	proc, err := procfs.NewDefaultFS()
+	if err != nil {
+		return errors.Wrap(err, "network metric: failed to obtain proc object")
+	}
+	netDevs, err := proc.NetDev()
+	if err != nil {
+		return errors.Wrap(err, "network metric: failed to obtain netDevs object")
+	}
+
+	rxBytes, rxPackets, txBytes, txPackets := uint64(0), uint64(0), uint64(0), uint64(0)
+	for _, netDev := range netDevs {
+		rxBytes += netDev.RxBytes
+		rxPackets += netDev.RxPackets
+		txBytes += netDev.TxBytes
+		txPackets += netDev.TxPackets
+	}
+	n.ReceiveBytesGauge.Set(float64(rxBytes))
```

Comment:
Using `WithLabelValues` you can report the values for each interface separately. (same for other network metrics).